### ISSUE TITLE
Reparenting an agent sets its HSM to inactive

### ISSUE THREAD

**Issue by andmish** [limbonaut/limboai#128]
Hello!
In running project (**Godot 4.2.2 | LimboAI 1.0.2**), reparenting an agent (that has **LimboHSM**) will set that **HSM** to inactive.
It was working correctly in **1.0.1**. So something happened between **1.0.1** >> **1.0.2**

Let's say we have such scene:
![2024-06-02 13_53_52](https://github.com/limbonaut/limboai/assets/37307893/1463e159-195e-42df-a3d1-63dba36289bb)

**Agent** has such script:
```GDScript
extends Node3D
@onready var states: LimboHSM = $States
@onready var move: LimboHSM = $States/Move
@onready var new_parent = $"../NewParent"

func _ready():	
	#states.initial_state = move

	states.initialize(self)
	states.set_active(true)
	
	await get_tree().create_timer(2.0).timeout
	reparent(new_parent)
	print(states.is_active())
	### prints "false"
```
After 2 seconds HSM will be inactive. Basically, reparenting disables AIs 🙂
 
<br>

**Other minor issue related to LimboHSM**

**World** has such script for reloading:
```GDScript
func _input(event: InputEvent) -> void:
	if event is InputEventKey and event.pressed:
		if event.keycode == KEY_R:
			get_tree().reload_current_scene()
```
Now if we alter **Agent**'s script (so initial state now is another **LimboHSM** child):
```GDScript
func _ready():	
	states.initial_state = move

	states.initialize(self)
	states.set_active(true)
	
	#await get_tree().create_timer(2.0).timeout
	#reparent(new_parent)
	#print(states.is_active())
	### prints "false"
```
and press **R** to reload the scene, we will get this error:
```
E 0:00:03:0596   LimboHSM::_exit: Condition "active_state == nullptr" is true.
  <C++ Source>   limboai\hsm\limbo_hsm.cpp:73 @ LimboHSM::_exit()
```
It happens only if the current substate is **LimboHSM** (not **LimboState**). So if the agent's current state is **Idle** reloading won't print an error.
Now if we uncomment everything:
```GDScript
func _ready():	
	states.initial_state = move

	states.initialize(self)
	states.set_active(true)
	
	await get_tree().create_timer(2.0).timeout
	reparent(new_parent)
	print(states.is_active())
```
it makes it inactive and also prints this error again (15 line is `reparent(new_parent)`):
```
E 0:00:04:0100   test_hsm.tscn::GDScript_gd7pe:15 @ _ready(): Condition "active_state == nullptr" is true.
  <C++ Source>   limboai\hsm\limbo_hsm.cpp:73 @ LimboHSM::_exit()
  <Stack Trace>  test_hsm.tscn::GDScript_gd7pe:15 @ _ready()
```

**Comment by limbonaut** [comment]
The reason why it happens is these lines: https://github.com/limbonaut/limboai/blob/2c8e0d2da08184051b2729b2eb6c977ab76027f2/hsm/limbo_state.cpp#L188C1-L192C11

It is done in order for a state or a behavior tree in a `BTState` to release resources they may acquire on `_enter`.  Like an instance of a pathfinder object, which a game may provide a limited number of. However, this is done with the intention to release resources when a state node is removed from a scene, and in your case it's reparenting that triggers this code path.

**Comment by limbonaut** [comment]
The second issue, I can see what happens. When `NOTIFICATION_EXIT_TREE` is sent, first is the inner HSM that receives this notification, and then the root HSM, which tries to exit its active state, but it's no longer active, so it trips the sanity check and prints the error message.

**Comment by limbonaut** [comment]
@andmish 
I opened PR that should fix both issues. Could you check if this build fixes those issues for you? <https://github.com/limbonaut/limboai/actions/runs/9350197568?pr=131>

**Comment by andmish** [comment]
@limbonaut I was checking notifications and found `NOTIFICATION_PREDELETE` that goes before exits, so I intended to ask you about it, but looks like you already know everything 😅

I've tested it, no issues, after reparenting an agent its hsm is still active and no error on reloading scene now. Big thanks! 🤗

**Comment by telemooon** [comment]
Thank you both so so much! I was having this exact issue and was trying to get around it by repeatedly calling hsm.set_active(true). This definitely saved a lot of headache.

**Comment by limbonaut** [comment]
I had to revert the fix in #226 due to a potential crash on exit in LimboHSM. I implemented an alternative solution in that PR, but it's not ideal.
